modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java

```diff
@@ -1028,4 +1087,357 @@ public Set<ReplicationGroupId> startedGroups() {
                 .map(Entry::getKey)
                 .collect(toSet());
     }
+
+    @TestOnly
+    public boolean isReplicaPrimaryOnly(ReplicationGroupId groupId) {
+        return replicaLifecycle.isReplicaPrimaryOnly(groupId);
+    }
+
+    private static class ReplicaLifecycle {
+        private final IgniteLogger log = Loggers.forClass(ReplicaLifecycle.class);
+
+        final Map<ReplicationGroupId, ReplicaLifecycleContext> replicaContexts = new ConcurrentHashMap<>();
+
+        final Executor replicaStartStopPool;
+
+        final ClockService clockService;
+
+        final PlacementDriver placementDriver;
+
+        final ReplicaManager replicaManager;
+
+        String localNodeId;
+
+        ReplicaLifecycle(Executor replicaStartStopPool, ClockService clockService, PlacementDriver placementDriver,
+                ReplicaManager replicaManager) {
+            this.replicaStartStopPool = replicaStartStopPool;
+            this.clockService = clockService;
+            this.placementDriver = placementDriver;
+            this.replicaManager = replicaManager;
+        }
+
+        void start(String localNodeId) {
+            this.localNodeId = localNodeId;
+            placementDriver.listen(PrimaryReplicaEvent.PRIMARY_REPLICA_ELECTED, this::onPrimaryElected);
+            placementDriver.listen(PrimaryReplicaEvent.PRIMARY_REPLICA_EXPIRED, this::onPrimaryExpired);
+        }
+
+        private CompletableFuture<Boolean> onPrimaryElected(PrimaryReplicaEventParameters parameters) {
+            ReplicaLifecycleContext context = getContext(parameters.groupId());
+
+            synchronized (context) {
+                if (localNodeId.equals(parameters.leaseholderId())) {
+                    assert context.replicaState != ReplicaState.STOPPED : "Unexpected primary replica state STOPPED [groupId="
+                            + parameters.groupId() + "].";
+
+                    context.reservedForPrimary = true;
+                } else {
+                    context.reservedForPrimary = false;
```

Comment:
BTW, why reserved for primary is not the state?

Reply:
I've thought about it. It would require the split of ASSIGNED state to something like ASSIGNED and ASSIGNED_PRIMARY, this would complicate the transition between states and adding the state transition logic to #onPrimaryElected, #onPrimaryExpired and #reserveReplica method. Now the replica states reflect it's lifecycle, and `reservedForPrimary` exists in parallel with that, it seems to be more simple and intuitive.